### DISCUSSION TITLE
Select-combobox keyboard navigation

### DISCUSSION THREAD

**Discussion by marcospassos** [ariakit/ariakit#1451] [Questions] [answered]
I'm working on a hybrid select-combobox component based on the [official example](https://ariakit.org/examples/select-combobox). Our goal is to allow toggling search capability and choosing between displaying the input inside or outside the popover through props.

Based on the provided example, I could make it work with very little trouble. I just had to replace the `<SelectPopover>` with the `<ComboboxPopover>` when rendering the input outside. However, the navigation isn't working well without the input: I've to press the arrow twice to go to the following item on the first and last items. I believe this is happening because, when using the combobox, this is the element that gets focused.

Is there any way to specify that the navigation should have the same behavior as the standalone select (loop keyboard navigation + select the first item of the list when opening) using the same wiring as the select-combobox example?

**Comment by diegohaz** [comment, marked as answer]
~I don't know how to reproduce what you described. Could you please create a CodeSandbox or something?~

Anyway, if you want to render the combobox input outside, you can just follow https://ariakit.org/examples/combobox

Edit: Okay, the select context confused me.

- Exclude the combobox input from the arrow key navigation:

  ```jsx
  useComboboxState({ includesBaseElement: false })
  ```

- Auto-select the first item (when typing)

  ```jsx
  <Combobox autoSelect />
  ```

- Auto-select the first item (always)

  ```jsx
  const combobox = useComboboxState()
  const { activeId, move, first } = combobox
  
  useEffect(() => {
    if (activeId === null) {
      move(first())
    }
  }, [activeId, move, first]);
  ```
  
  We can possibly add an `autoSelect="always"` prop to the combobox component. If you think that would be helpful, please submit a feature request so we can track it and see if more people have the same need.

**Reply by marcospassos** [reply]
That solved the issue, thank you!

**Reply by gylxan** [reply]
Exactly what we searched for!
We just switched from [cmdk](https://github.com/pacocoursey/cmdk) to ariakit and were missing the feature that the first option is default highlighted. 

Unfortunately the answer seems a little bit outdated.  @diegohaz Can you tell us, how we would implement it with the todays hook `useComboboxContext` and/or `useComboboxState`?

I currently have an implementation like this, but wondered if there is a better way.
```tsx
  const comboboxStore = useComboboxStore({
    includesBaseElement: false,
  })
  const { move, first } = comboboxStore
  const comboboxState = comboboxStore.useState()
  const { activeId } = comboboxState
  const firstItem = first()
  useEffect(() => {
    if (activeId === null) {
      move(firstItem)
    }
  }, [activeId, move, firstItem])
```

I think, I will also create a feature request for your suggestion of adding `autoSelect="always"` prop to the combobox.

**Reply by diegohaz** [reply]
@gylxan You can use the Dialog's [`initialFocus`](https://ariakit.org/reference/dialog#initialfocus) prop. Pass a ref to the first item:

```jsx
const initialFocusRef = useRef(null);

<Dialog initialFocus={initialFocusRef}>
  <ComboboxProvider open>
    <Combobox />
    <ComboboxList>
      {items.map((value, index) => (
        <ComboboxItem key={value} ref={index === 0 ? initialFocusRef : null}>
```

[`autoFocus`](https://ariakit.org/reference/combobox-item#autofocus) should also work:

```jsx
      {items.map((value, index) => (
        <ComboboxItem key={value} autoFocus={index === 0}>
```

**Reply by gylxan** [reply]
We don't use a Dialog, but the `autoFocus` property works like a charm! Just had to find the index of the first non-disabled item.

Thanks @diegohaz for your great quick help!

**Reply by diegohaz** [reply]
The `autoSelect="always"` feature has been implemented as of [`v0.4.1`](https://ariakit.org/changelog#041).

**Reply by gylxan** [reply]
Wow, nice! Just tested and it works perfectly! Thanks again @diegohaz ❤️!
